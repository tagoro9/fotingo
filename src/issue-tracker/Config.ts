export interface JiraConfig {
  user: {
    login: string;
    token: string;
  };
  releaseTemplate: string;
  root: string;
}
