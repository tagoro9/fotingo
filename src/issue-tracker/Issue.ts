/**
 * Issue
 */

export enum IssueType {
  BUG = 'Bug',
  FEATURE = 'Feature',
  STORY = 'Story',
  SUB_TASK = 'Sub-task',
  TASK = 'Task',
}

export interface IssueEditMeta {
  fields: object;
}

export interface IssueCommentAuthor {
  name: string;
  displayName: string;
  active: boolean;
}

export interface IssueComment {
  id: string;
  author: IssueCommentAuthor;
  updateAuthor: IssueCommentAuthor;
  body: string;
  updated: string;
  created: string;
}

export interface IssueTypeData {
  id: number;
  name: string;
  shortName: string;
}

export enum IssueStatus {
  BACKLOG = 'BACKLOG',
  DONE = 'DONE',
  IN_PROGRESS = 'IN_PROGRESS',
  IN_REVIEW = 'IN_REVIEW',
  SELECTED_FOR_DEVELOPMENT = 'SELECTED_FOR_DEVELOPMENT',
}

export interface IssueTransition {
  id: number;
  name: string;
}

// TODO Decouple this from the jira specific fields. There shoudl be a JiraIssue interface and a mapper from that one to this one
export interface Issue {
  id: number;
  key: string;
  sanitizedSummary: string;
  type: IssueType;
  transitions: IssueTransition[];
  renderedFields: {
    description?: string;
  };
  fields: {
    summary: string;
    description?: string;
    issuetype: {
      name: string;
    };
    project: {
      id: string;
    };
  };
  url: string;
}

export interface CreateIssue {
  description?: string;
  labels?: string[];
  project: string;
  title: string;
  type: IssueType;
}

export interface CreateRelease {
  name: string;
  issues: Issue[];
  submitRelease: boolean;
  useDefaults: boolean;
}

export interface GetIssue {
  id: string;
}

export interface User {
  key: string;
  groups: {
    items: {
      name: string;
    };
  };
}

export interface Project {
  id: number;
  issueTypes: { [type in IssueType]: IssueTypeData };
  name: string;
  key: string;
}

export interface JiraProject {
  id: number;
  issueTypes: Array<{ id: number; name: string }>;
  name: string;
  key: string;
}

export interface ReleaseNotes {
  title: string;
  body: string;
}

export interface Release {
  id: string;
  name: string;
  issues: Issue[];
  url?: string;
}

export interface JiraRelease {
  self: string;
  id: string;
  description: string;
  name: string;
  archived: boolean;
  released: boolean;
  releaseDate: string;
  userReleaseDate: string;
  project: Project;
  projectId: number;
}
