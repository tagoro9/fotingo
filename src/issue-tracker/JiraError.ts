export interface JiraError {
  readonly code: string | number;
  readonly message: string;
}

export class JiraErrorImpl extends Error implements JiraError {
  public readonly code: string | number;

  constructor(message: string, code: string | number) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
    this.code = code;
  }
}
