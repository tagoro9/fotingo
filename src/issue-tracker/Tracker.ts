import { Observable } from 'rxjs';
import {
  CreateIssue,
  CreateRelease,
  Issue,
  IssueComment,
  IssueStatus,
  Release,
  User,
} from 'src/types';

export interface Tracker {
  getIssue: (issueId: string) => Observable<Issue>;
  createIssue: (data: CreateIssue, user: User) => Observable<Issue>;
  createIssueForCurrentUser: (data: CreateIssue) => Observable<Issue>;
  getCurrentUser: () => Observable<User>;
  setIssueStatus: (status: IssueStatus, issueId: string) => Observable<Issue>;
  addCommentToIssue: (issueId: string, comment: string) => Observable<IssueComment>;
  addLabelToIssue: (issueId: string, label: string) => Observable<Issue>;
  createRelease: (data: CreateRelease) => Observable<Release>;
  isValidIssueName: (name: string) => boolean;
  setIssuesFixVersion(release: Release): Observable<Release>;
}
