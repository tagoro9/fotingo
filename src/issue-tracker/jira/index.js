import R from 'ramda';
import httpClient from '../http-client';
import {
  debug,
  debugCurried,
  debugCurriedP,
  wrapInPromise,
  allowUserToEditMessage,
} from '../../util';
import reporter from '../../reporter';
import { errors, catchPromiseAndThrow, throwControlledError } from '../../error';
import issueUtil from './issue';
import readIssueStatus, { status } from './status';

export default config => () => {
  debug('jira', 'Initializing Jira api');
  const root = config.get(['jira', 'root'])
    ? wrapInPromise(config.get(['jira', 'root']))
    : R.composeP(config.update(['jira', 'root']), reporter.question)({
        question: "What's your jira root?",
      });

  return root.then(jiraRoot => {
    const { get, post, put, setAuth } = httpClient(jiraRoot);
    const issueRoot = `${jiraRoot}/browse/`;
    const readUserInfo = () => {
      // It doesn't feel like logger should be used here
      debug('jira', 'Reading user login info');
      const readUsernamePromise = reporter.question({ question: "What's your Jira username?" });
      const readPasswordPromise = readUsernamePromise.then(
        R.partial(reporter.question, [{ question: "What's your Jira password?", password: true }]),
      );
      return Promise.all([readUsernamePromise, readPasswordPromise]).then(([login, password]) => ({
        login,
        password,
      }));
    };

    const getCurrentUser = () => get('/rest/api/2/myself?expand=groups').then(R.prop('body'));

    const getIssueTypes = () => get('/rest/api/2/issuetype').then(R.prop('body'));

    const getShortName = name => {
      if (name.match(/feature|story/i)) {
        return 'f';
      }
      if (name.match(/task/i)) {
        return 'c';
      }
      return name[0].toLowerCase();
    };

    const initialize = user =>
      R.composeP(
        R.always(user),
        R.compose(
          R.ifElse(
            R.compose(R.either(R.isNil, R.isEmpty)),
            R.composeP(
              R.compose(
                wrapInPromise,
                config.update(['jira', 'issueTypes']),
                R.mergeAll,
                R.map(({ id, name }) => ({ [id]: { name, shortName: getShortName(name) } })),
              ),
              getIssueTypes,
            ),
            R.compose(wrapInPromise, R.identity),
          ),
          config.get,
        ),
      )(['jira', 'issueTypes']);

    const doLogin = R.composeP(
      R.compose(catchPromiseAndThrow('jira', errors.jira.couldNotAuthenticate), getCurrentUser),
      setAuth,
      config.update(['jira', 'user']),
      readUserInfo,
    );
    let loginPromise;
    if (config.isJiraLoggedIn()) {
      setAuth(config.get(['jira', 'user']));
      loginPromise = getCurrentUser().catch(
        R.compose(doLogin, debugCurried('jira', 'Current authentication failed. Attempting login')),
      );
    } else {
      loginPromise = doLogin();
    }

    const parseIssue = R.compose(
      wrapInPromise,
      R.converge(R.set(R.lensProp('url')), [
        R.compose(R.concat(issueRoot), R.prop('key')),
        R.identity,
      ]),
      R.prop('body'),
    );

    const addCommentToIssue = R.curry((issue, comment) =>
      post(`/rest/api/2/issue/${issue.key}/comment`, { body: { body: comment } }),
    );

    return loginPromise.then(initialize).then(user => {
      const jira = {
        name: 'jira',
        issueRoot,
        getCurrentUser: R.always(wrapInPromise(user)),
        getIssue: R.composeP(
          parseIssue,
          R.compose(
            catchPromiseAndThrow('jira', errors.jira.issueNotFound),
            get,
            R.concat(R.__, '?expand=transitions'),
            R.concat('/rest/api/2/issue/'),
          ),
          debugCurriedP('jira', 'Getting issue from jira'),
        ),
        setIssueStatus: R.curryN(2, ({ status: issueStatus, comment }, issue) =>
          R.composeP(
            R.always(issue),
            // Jira api for transition is not adding the comment so we need an extra api call
            R.partial(R.unless(R.isNil, addCommentToIssue(issue)), [comment]),
            post(`/rest/api/2/issue/${issue.key}/transitions`),
            debugCurriedP('jira', `Updating issue status to ${issueStatus}`),
            statuses =>
              wrapInPromise({
                body: {
                  transition: R.compose(
                    R.pick(['id']),
                    R.find(
                      R.compose(R.equals(statuses[issueStatus]), Number, R.path(['to', 'id'])),
                    ),
                    R.prop('transitions'),
                  )(issue),
                  fields: {},
                },
              }),
            readIssueStatus(config),
          )(issue),
        ),
        canWorkOnIssue: R.curryN(
          2,
          R.converge((canWonOnIssue, promise) => promise.then(canWonOnIssue), [
            R.curryN(2, R.invoker(2, 'canWorkOnIssue')),
            R.composeP(R.compose(wrapInPromise, issueUtil), R.flip(readIssueStatus(config))),
          ]),
        ),
        // String -> Array -> String
        createReleaseNotes: R.curryN(
          2,
          R.composeP(
            R.ifElse(R.isEmpty, throwControlledError(errors.jira.releaseNotesInvalid), notes => ({
              title: R.compose(R.head, R.split('\n'), R.trim)(notes),
              body: R.compose(
                R.join('\n'),
                R.filter(R.compose(R.not, R.isEmpty)),
                R.tail,
                R.split('\n'),
                R.trim,
              )(notes),
            })),
            allowUserToEditMessage(`/tmp/fotingo-notes-file-${Date.now()}`),
            R.compose(
              wrapInPromise,
              R.converge(R.unapply(R.join('\n')), [
                R.compose(
                  R.concat(R.__, '\n'),
                  R.ifElse(
                    R.compose(R.equals(1), R.length, R.nthArg(1)),
                    R.compose(R.view(R.lensPath(['fields', 'summary'])), R.head, R.nthArg(1)),
                    R.nthArg(0),
                  ),
                ),
                R.compose(
                  R.join('\n'),
                  R.map(issue => `* [${issue.key}](${issue.url}). ${issue.fields.summary}.`),
                  R.nthArg(1),
                ),
              ]),
            ),
          ),
        ),
        createVersion: R.curryN(2, (releaseId, issues) =>
          R.composeP(
            R.prop('body'),
            post('/rest/api/2/version'),
            R.set(R.lensPath(['body', 'projectId']), R.__, {
              body: {
                archived: false,
                released: true,
                releaseDate: new Date().toISOString().substring(0, 10),
                name: releaseId,
                description: 'Release from fotingo',
              },
            }),
            R.head,
            R.when(
              R.compose(R.lt(1), R.length, R.uniq),
              throwControlledError(errors.jira.issuesInMultipleProjects),
            ),
            R.map(R.view(R.lensPath(['fields', 'project', 'id']))),
            wrapInPromise,
          )(issues),
        ),
        setIssuesFixVersion: R.curryN(2, (issues, version) => {
          return R.compose(
            promises => Promise.all(promises),
            R.map(
              R.converge((...promises) => Promise.all(promises), [
                R.compose(
                  put(R.__, {
                    body: {
                      update: {
                        fixVersions: [
                          {
                            add: { name: version.name },
                          },
                        ],
                      },
                    },
                  }),
                  R.concat('/rest/api/2/issue/'),
                  R.prop('key'),
                ),
                jira.setIssueStatus({ status: status.DONE }),
              ]),
            ),
          )(issues);
        }),
        status,
      };
      return jira;
    });
  });
};
