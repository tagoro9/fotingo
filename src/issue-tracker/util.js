import R from 'ramda';
import { throwControlledError, errors } from '../error';

const isPresent = R.complement(R.either(R.isNil, R.isEmpty));
const issueRegex = new RegExp('^([a-zA-Z]+)\\-(\\d+)$');
const isValidName = R.test(issueRegex);
const isInvalid = R.complement(R.both(isPresent, isValidName));
// string -> string
export const validateIssueId = R.when(isInvalid, throwControlledError(errors.jira.issueIdNotValid));
// string -> string
export const validateIssueDescription = R.when(
  R.complement(isPresent),
  throwControlledError(errors.jira.issueDescriptionNotValid),
);
