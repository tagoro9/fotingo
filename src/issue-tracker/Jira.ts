import { boundMethod } from 'autobind-decorator';
import {
  always,
  compose,
  converge,
  // curry,
  evolve,
  groupBy,
  head,
  indexBy,
  join,
  length,
  lensProp,
  lt,
  map as rMap,
  mapObjIndexed,
  path,
  prop,
  replace,
  split,
  tail,
  take,
  toLower,
  toPairs,
  trim,
  unapply,
  uniq,
  view,
  when,
  zipObj,
} from 'ramda';
import { from, merge, Observable, of, throwError } from 'rxjs';
import { catchError, map, reduce, switchMap } from 'rxjs/operators';
import { editVirtualFile } from 'src/io/file-util';
import { Messenger } from 'src/io/messenger';
import { parseTemplate } from 'src/io/template-util';
import { HttpClient } from 'src/network/HttpClient';
import { HttpError } from 'src/network/HttpError';
import * as Turndown from 'turndown';

import { JiraConfig } from './Config';
import {
  CreateIssue,
  CreateRelease,
  Issue,
  IssueComment,
  IssueEditMeta,
  IssueStatus,
  IssueTransition,
  IssueType,
  IssueTypeData,
  JiraProject,
  JiraRelease,
  Project,
  Release,
  ReleaseNotes,
  User,
} from './Issue';
import { JiraErrorImpl } from './JiraError';
import { Tracker } from './Tracker';

const turnDownService = new Turndown();

const getShortName = (name: string) => {
  if (name.match(/feature|story/i)) {
    return 'f';
  }
  if (name.match(/task/i)) {
    return 'c';
  }
  return name[0].toLowerCase();
};

// TODO Make this part of config
const statusRegex = {
  [IssueStatus.BACKLOG]: /backlog/i,
  [IssueStatus.IN_PROGRESS]: /in progress/i,
  [IssueStatus.IN_REVIEW]: /review/i,
  [IssueStatus.DONE]: /done/i,
  [IssueStatus.SELECTED_FOR_DEVELOPMENT]: /(todo)|(to do)|(selected for development)/i,
};

enum RELEASE_TEMPLATE_KEYS {
  VERSION = 'version',
  FIXED_ISSUES_BY_CATEGORY = 'fixedIssuesByCategory',
  FOTINGO_BANNER = 'fotingo.banner',
}

const ISSUE_TYPE_TO_RELEASE_SECTION: { [k in IssueType]: string } = {
  [IssueType.TASK]: 'Features',
  [IssueType.SUB_TASK]: 'Features',
  [IssueType.BUG]: 'Bug fixes',
  [IssueType.STORY]: 'Features',
  [IssueType.FEATURE]: 'Features',
};

// Using Jira REST API v2 (https://developer.atlassian.com/cloud/jira/platform/rest/v2)
export class Jira implements Tracker {
  private client: HttpClient;
  private config: JiraConfig;
  private messenger: Messenger;

  constructor(config: JiraConfig, messenger: Messenger) {
    this.config = config;
    this.messenger = messenger;
    // TODO Add a new client for agile API https://developer.atlassian.com/cloud/jira/software/rest/
    this.client = new HttpClient({
      allowConcurrentRequests: false,
      auth: { pass: config.user.token, user: config.user.login },
      root: `${config.root}/rest/api/2`,
    });
  }

  public setIssueStatus(status: IssueStatus, issueId: string): Observable<Issue> {
    return this.getIssue(issueId).pipe(
      switchMap(issue => {
        const transition = this.getTransitionForStatus(issue, status);
        if (!transition) {
          throw new Error('Missing status');
        }
        return this.client
          .post<Issue>(`/issue/${issue.key}/transitions`, {
            body: {
              transition: transition.id,
            },
          })
          .pipe(map(always(issue)));
      }),
      catchError(this.mapError),
    );
  }

  public getCurrentUser(): Observable<User> {
    return this.client
      .get<User>('/myself', {
        qs: { expand: 'groups' },
      })
      .pipe(
        map(prop('body')),
        catchError(this.mapError),
      );
  }

  @boundMethod
  public getIssue(id: string): Observable<Issue> {
    return this.client
      .get<Issue>(`/issue/${id}`, {
        qs: { expand: 'transitions, renderedFields' },
      })
      .pipe(
        map(prop('body')),
        map(issue => ({
          ...issue,
          fields: {
            ...issue.fields,
            description: issue.renderedFields.description
              ? turnDownService.turndown(issue.renderedFields.description)
              : undefined,
          },
          sanitizedSummary: compose<string, string, string, string, string, string, string>(
            take(72),
            replace(/(_|-)$/, ''),
            replace(/\s|\(|\)|__+/g, '_'),
            replace(/\/|\.|--=/g, '-'),
            replace(/,|\[|]|"|'|”|“|@|’|`|:|\$|\?|\*|<|>|&|~/g, ''),
            toLower,
          )(issue.fields.summary),
          type: issue.fields.issuetype.name as IssueType,
          url: `${this.config.root}/browse/${issue.key}`,
        })),
        catchError(this.mapError),
      );
  }

  public isValidIssueName(name: string): boolean {
    return /\w+-\d+/i.test(name);
  }

  public getProject(id: string): Observable<Project> {
    return this.client.get<JiraProject>(`/project/${id}`).pipe(
      map(
        compose(
          (data: JiraProject) =>
            evolve(
              {
                issueTypes: compose(
                  indexBy<IssueTypeData>(prop('name')),
                  rMap(
                    (typeData: { id: number; name: string }): IssueTypeData => ({
                      ...typeData,
                      shortName: getShortName(typeData.name),
                    }),
                  ),
                ) as (d: JiraProject['issueTypes']) => Project['issueTypes'],
              },
              data,
            ),
          prop('body'),
        ),
      ),
      catchError(this.mapError),
    );
  }

  @boundMethod
  public createIssueForCurrentUser(data: CreateIssue): Observable<Issue> {
    return this.getCurrentUser().pipe(switchMap(user => this.createIssue(data, user)));
  }

  @boundMethod
  public createIssue(data: CreateIssue, user: User): Observable<Issue> {
    return this.getProject(data.project).pipe(
      switchMap(project =>
        this.client
          .post<Issue>('/issue', {
            body: {
              fields: {
                assignee: {
                  name: user.key,
                },
                description: data.description,
                issuetype: {
                  id: project.issueTypes[data.type].id,
                },
                // TODO Make sure labels exist
                labels: data.labels,
                project: {
                  id: project.id,
                },
                summary: data.title,
              },
            },
          })
          .pipe(
            map(prop('body')),
            catchError(this.mapError),
          ),
      ),
    );
  }

  public addLabelToIssue(issueId: string, label: string): Observable<Issue> {
    this.messenger.emit('This is a test so TS compiles');
    return this.client
      .put<void>(`/issue/${issueId}/`, {
        body: {
          update: {
            labels: [{ add: label }],
          },
        },
      })
      .pipe(
        catchError(this.mapError),
        switchMap(() => this.getIssue(issueId)),
      );
  }

  public getIssueEditMeta(issueId: string): Observable<IssueEditMeta> {
    return this.client.get<IssueEditMeta>(`/issue/${issueId}/editmeta`).pipe(
      map(prop('body')),
      catchError(this.mapError),
    );
  }

  public addCommentToIssue(issueId: string, comment: string): Observable<IssueComment> {
    return this.getIssue(issueId).pipe(
      switchMap(issue =>
        this.client.post<IssueComment>(`/issue/${issue.key}/comment`, {
          body: { body: comment },
        }),
      ),
      map(prop('body')),
      catchError(this.mapError),
    );
  }

  @boundMethod
  public createRelease(data: CreateRelease): Observable<Release> {
    return of(data).pipe(
      map(
        when(
          compose(
            lt(1),
            length,
            uniq,
            rMap(path(['fields', 'project', 'id'])),
            prop('issues'),
          ),
          () => {
            throw new Error('Issues in multiple projects');
          },
        ),
      ),
      switchMap(this.createReleaseNotes),
      map(
        converge(unapply(zipObj(['title', 'body'])), [
          compose<string, string, string, string[], string>(
            head,
            split('\n'),
            trim,
          ),
          compose<string, string, string, string[], string[], string>(
            join('\n'),
            tail,
            split('\n'),
            trim,
          ),
        ]),
      ),
      switchMap(notes => {
        if (data.submitRelease) {
          return this.createVersion(data, notes).pipe(
            map(release => ({
              id: release.id,
              issues: data.issues,
              name: release.name,
              notes,
              url: `${this.config.root}/projects/${release.project.key}/versions/${release.id}`,
            })),
          );
        }
        return of({
          id: data.name,
          issues: data.issues,
          name: data.name,
          notes,
        });
      }),
    );
  }

  @boundMethod
  public setIssuesFixVersion(release: Release): Observable<Release> {
    return from(release.issues)
      .pipe(
        switchMap(issue =>
          merge(
            this.setIssueStatus(IssueStatus.DONE, issue.key),
            this.client
              .put<Issue>(`/issue/${issue.key}`, {
                body: {
                  update: {
                    fixVersions: [{ add: { name: release.name } }],
                  },
                },
              })
              .pipe(map(prop('body'))),
          ).pipe(
            reduce<Issue, Issue[]>((acc, val) => acc.concat(val), []),
            map(issues => head(issues)),
          ),
        ),
      )
      .pipe(
        reduce<Issue, Issue[]>((acc, val) => acc.concat(val), []),
        map(always(release)),
      );
  }

  @boundMethod
  private createReleaseNotes(data: CreateRelease): Observable<string> {
    const initialReleaseNotes = this.getReleaseNotesFromTemplate(data);
    return from(
      data.useDefaults ? [initialReleaseNotes] : this.editReleaseNotes(initialReleaseNotes),
    );
  }

  private async editReleaseNotes(initialReleaseNotes: string): Promise<string> {
    this.messenger.inThread(true);
    const notes = await editVirtualFile({
      extension: 'md',
      initialContent: initialReleaseNotes,
      prefix: 'fotingo-review',
    });
    this.messenger.inThread(false);
    return notes;
  }

  private getReleaseNotesFromTemplate(data: CreateRelease): string {
    return parseTemplate<RELEASE_TEMPLATE_KEYS>({
      data: {
        [RELEASE_TEMPLATE_KEYS.FIXED_ISSUES_BY_CATEGORY]: compose(
          join('\n'),
          rMap(([title, list]) => `# ${title}:\n\n${list}`),
          toPairs,
          mapObjIndexed(
            compose(
              join('\n'),
              rMap((issue: Issue) => `* [#${issue.key}](${issue.url}): ${issue.fields.summary}`),
            ),
          ),
          groupBy(
            compose(
              lens => view(lens, ISSUE_TYPE_TO_RELEASE_SECTION),
              lensProp,
              prop('type'),
            ),
          ),
        )(data.issues),
        [RELEASE_TEMPLATE_KEYS.VERSION]: data.name,
        [RELEASE_TEMPLATE_KEYS.FOTINGO_BANNER]:
          '🚀 Release created with [fotingo](https://github.com/tagoro9/fotingo)',
      },
      template: this.config.releaseTemplate,
    });
  }

  private createVersion(data: CreateRelease, releaseNotes: ReleaseNotes): Observable<JiraRelease> {
    return this.client
      .post<JiraRelease>(`/version`, {
        body: {
          archived: false,
          description: releaseNotes.title,
          name: data.name,
          projectId: head(data.issues.map(path(['fields', 'project', 'id']))),
          releaseDate: new Date().toISOString().substring(0, 10),
          released: true,
        },
      })
      .pipe(
        map(prop('body')),
        switchMap(release =>
          this.getProject(String(release.projectId)).pipe(
            map(project => ({ ...release, project })),
          ),
        ),
      );
  }

  private getTransitionForStatus(issue: Issue, status: IssueStatus): IssueTransition | undefined {
    return issue.transitions.find(transition => statusRegex[status].test(transition.name));
  }

  private mapError(e: NodeJS.ErrnoException | HttpError) {
    if ('status' in e) {
      const code = e.status;
      const message =
        (e.body.errorMessages && e.body.errorMessages[0]) ||
        'Something went wrong when connecting with jira';
      return throwError(new JiraErrorImpl(message, code));
    } else {
      return throwError(e);
    }
  }
}
