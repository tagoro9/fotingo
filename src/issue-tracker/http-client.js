import request from 'request';
import R from 'ramda';
import { debug } from '../util';

const handleServerResponse = (resolve, reject) =>
  R.ifElse(
    R.compose(R.not, R.isNil, R.nthArg(0)),
    reject,
    R.ifElse(
      R.compose(R.propSatisfies(R.gt(400), 'statusCode'), R.nthArg(1)),
      R.unapply(R.compose(resolve, args => ({ response: args[1], body: args[2] }))),
      R.compose(
        reject,
        R.unapply(args => {
          R.compose(
            debug('http'),
            R.concat('Request failed with status code '),
            R.prop('statusCode'),
            R.nth(1),
          )(args);
          return args[2];
        }),
      ),
    ),
  );

const headers = { accept: 'application/json' };

export default function(rootUrl) {
  const makeUrl = R.concat(rootUrl);
  let auth = {};
  const jar = request.jar();
  const serverCall = R.curry(
    (method, url, { form, body }) =>
      new Promise((resolve, reject) => {
        debug(
          'http',
          `Performing ${method} ${makeUrl(url)} ${body
            ? `with body ${JSON.stringify(body, null, 2)}`
            : ''}`,
        );
        return request(
          { url: makeUrl(url), body, form, headers, jar, json: true, method, auth },
          handleServerResponse(resolve, reject),
        );
      }),
  );
  const setCookieToJar = R.compose(
    R.bind(R.partialRight(jar.setCookie, [rootUrl]), jar),
    request.cookie,
  );

  return {
    post: serverCall('POST'),
    get: serverCall('GET', R.__, {}),
    setAuth: ({ login, password }) => {
      auth = { ...auth, user: login, pass: password };
    },
    setCookie: R.ifElse(R.is(Array), R.map(setCookieToJar), setCookieToJar),
  };
}
