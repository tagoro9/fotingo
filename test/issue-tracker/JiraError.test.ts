import 'jest';
import { JiraErrorImpl } from 'src/issue-tracker/JiraError';

describe('JiraError', () => {
  it('should store the message and the code', () => {
    const error = new JiraErrorImpl('Test message', 'ERR_CODE');
    expect(error.message).toMatchSnapshot();
    expect(error.code).toMatchSnapshot();
  });
});
