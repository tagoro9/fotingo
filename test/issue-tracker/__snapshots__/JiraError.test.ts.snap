// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JiraError should store the message and the code 1`] = `"Test message"`;

exports[`JiraError should store the message and the code 2`] = `"ERR_CODE"`;
